MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/OutputSurface.java

```diff
@@ -0,0 +1,288 @@
+     * @param surfaceHeight The height of the Surface object, in pixels.
+     */
+    @RequiresApi(api = Build.VERSION_CODES.KITKAT)
+    public void setSurface(Surface surface, int surfaceWidth, int surfaceHeight) {
+        // Set up the texture render
+        setupTextureRender(surfaceWidth, surfaceHeight);
+
+        renderHandler.post(() -> {
+            // Abort if context is null
+            if (eglContext == null) {
+                Log.i(TAG, "EGL Context is null, can't set EGL surface");
+                return;
+            }
+
+            // Destroy EGL surface if valid
+            if (!Objects.equals(eglSurface, EGL14.EGL_NO_SURFACE)) {
+                EGL14.eglDestroySurface(eglDisplay, eglSurface);
+            }
+
+            // Initialize EGL surface
+            if (surface == null) {
+                Log.d(TAG, "EGL initialized to no surface");
+                eglSurface = EGL14.EGL_NO_SURFACE;
+            } else {
+                Log.d(TAG, "Creating EGL surface");
+                eglSurface = EGL14.eglCreateWindowSurface(
+                        eglDisplay,
+                        eglConfig,
+                        surface,
+                        new int[] { EGL14.EGL_NONE },
+                        /* offset = */ 0
+                );
+            }
+
+            // Make EGL surface current
+            if (!EGL14.eglMakeCurrent(eglDisplay, eglSurface, eglSurface, eglContext)) {
+                Log.e(TAG, "Failed to make context current");
+            }
+        });
+    }
+
+    /**
+     * Free up all resources associated with this OutputSurface.
+     */
+    @RequiresApi(api = Build.VERSION_CODES.JELLY_BEAN_MR1)
+    public void release() {
+        Log.d(TAG, "Releasing output surface");
+        renderHandler.post(() -> {
+            if (eglDisplay != EGL14.EGL_NO_DISPLAY) {
+                EGL14.eglDestroySurface(eglDisplay, eglSurface);
+                EGL14.eglDestroyContext(eglDisplay, eglContext);
+                EGL14.eglReleaseThread();
+                EGL14.eglTerminate(eglDisplay);
+            }
+
+            eglDisplay = EGL14.EGL_NO_DISPLAY;
+            eglContext = EGL14.EGL_NO_CONTEXT;
+            eglSurface = EGL14.EGL_NO_SURFACE;
+
+            textureRender = null;
+            surfaceTexture.release();
+            surfaceTexture = null;
+            decodeSurface.release();
+            decodeSurface = null;
+        });
+    }
+
+    /**
+     * Get the Surface object which decoded video frames are rendered to before processing. Note
+     * that this Surface does not display the final rendered video, but instead holds the
+     * SurfaceTexture that supply textures to the TextureRender.
+     * @return the intermediate decoding Surface.
+     */
+    public Surface getDecodeSurface() {
+        SettableFuture<Surface> surfaceFuture = SettableFuture.create();
+        renderHandler.post(() -> surfaceFuture.set(decodeSurface));
```

Comment:
couldn't this get set with null ?

Reply:
If I synchronize the block where the decodeSurface is created, would that fix this? My intention is that this would only return null if getDecodeSurface() was called before setSurface() was called (where I initialize decodeSurface), and I would put a note of not to do in the method javadoc.